### PR TITLE
ceph-osd: fix autodetection activation

### PR DESCRIPTION
**Pull request by leseb**
Prior to this patch this activation sequence for autodetection was
always skipped because we were asking to activate on device without
partitions, which doesn't make sense.

We also fix the way we lookup for a device, since the data partition is
always numbered 1, we take the min element of the dict.

Closes: https://github.com/ceph/ceph-ansible/issues/1782
Signed-off-by: Sébastien Han <seb@redhat.com>

### CODE REVIEW

**Comment by leseb** [comment]
jenkins test luminous-ansible2.3-centos7_cluster

**Comment by leseb** [comment]
jenkins test luminous-ansible2.3-bluestore_cluster

**Comment by leseb** [comment]
jenkins test luminous-ansible2.3-docker_cluster

**Comment by leseb** [comment]
jenkins test luminous-ansible2.3-purge_docker_cluster

**Comment by leseb** [comment]
jenkins test luminous-ansible2.3-update_cluster

**Comment by leseb** [comment]
jenkins test luminous-ansible2.3-update_docker_cluster

**Comment by leseb** [comment]
lvm failures are expected, waiting for https://github.com/ceph/ceph/pull/17208
As for ansible2.3-bluestore_docker_dmcrypt_journal_collocation, the machine timed out due to a network issue. Safe to merge without those.

**Comment by leseb** [comment]
jenkins test this please

**Comment by leseb** [comment]
jenkins test luminous-ansible2.3-journal_collocation_auto

**Comment by leseb** [comment]
jenkins test luminous-ansible2.3-journal_collocation_auto

**Comment by leseb** [comment]
jenkins test luminous-ansible2.3-dmcrypt_journal_collocation

**Comment by leseb** [comment]
jenkins test luminous-ansible2.3-docker_cluster

**Comment by leseb** [comment]
jenkins test luminous-ansible2.3-journal_collocation_auto_dmcrypt

**Comment by leseb** [comment]
Somehow the CI sees 0 OSD:

```

self = <tests.mon.test_mons.TestOSDs object at 0x7f89b94b7610>
node = {'address': '192.168.3.10', 'ceph_stable_release': 'luminous', 'cluster_address': '', 'cluster_name': u'test', ...}
host = <testinfra.host.Host object at 0x7f89bac264d0>

    @pytest.mark.no_docker
    def test_all_osds_are_up_and_in(self, node, host):
        cmd = "sudo ceph --cluster={} --connect-timeout 5 -s".format(node["cluster_name"])
        output = host.check_output(cmd)
        phrase = "{num_osds} osds: {num_osds} up, {num_osds} in".format(num_osds=node["total_osds"])
>       assert phrase in output
E       AssertionError: assert '0 osds: 0 up, 0 in' in '  cluster:\n    id:     5f5555dd-448a-4477-b86b-91ccc508d722\n    health: HEALTH_WARN\n            no active mgr\n \n  ser...\n    pools:   0 pools, 0 pgs\n    objects: 0 objects, 0 bytes\n    usage:   0 kB used, 0 kB / 0 kB avail\n    pgs:     \n '
```

While reproducing on a slave I successfully get 3 OSDs...

File: roles/ceph-osd/tasks/scenarios/collocated.yml [hunk 1]
```diff
@@ -55,11 +55,9 @@
...
     - ceph_release_num.{{ ceph_release }} < ceph_release_num.luminous
 
-# NOTE (alahouze): if the device is a partition, the parted command below has
-# failed, this is why we check if the device is a partition too.
-- name: automatic prepare "{{ osd_objectstore }}" osd disk(s) without partitions with collocated osd data and journal
-  command: "ceph-disk prepare {{ ceph_disk_cli_options }} /dev/{{ item.key }}"
-  register: prepared_osds
+- name: generate device list when osd_auto_discovery
+  set_fact:
+    devices: "/dev/{{ item.key }}"
```
Comment (leseb) [inline roles/ceph-osd/tasks/scenarios/collocated.yml]: 👎 This won't work, this is not creating a list but just the last item of the list.

**Comment by leseb** [comment]
@andrewschoen any help with the CI issues will be appreciated. I think the patch works, we might need to rework the tests.

```
=================================== FAILURES ===================================
_____________ TestOSD.test_osds_are_all_collocated[ansible://osd0] _____________
[gw0] linux2 -- Python 2.7.13 /tmp/tox.HsMm8N7kan/luminous-ansible2.3-journal_collocation_auto/bin/python2

self = <tests.osd.test_journal_collocation.TestOSD object at 0x7f57aca55050>
node = {'address': '192.168.3.100', 'ceph_stable_release': 'luminous', 'cluster_address': '192.168.4.200', 'cluster_name': u'test', ...}
host = <testinfra.host.Host object at 0x7f57a7716490>

    def test_osds_are_all_collocated(self, node, host):
        # TODO: figure out way to paramaterize node['vars']['devices'] for this test
>       for device in node["vars"]["devices"]:
E       KeyError: 'devices'

../../../tests/osd/test_journal_collocation.py:6: KeyError
____________ TestOSDs.test_osd_services_are_running[ansible://osd0] ____________
[gw1] linux2 -- Python 2.7.13 /tmp/tox.HsMm8N7kan/luminous-ansible2.3-journal_collocation_auto/bin/python2

self = <tests.osd.test_osds.TestOSDs object at 0x7f75bc195ad0>
node = {'address': '192.168.3.100', 'ceph_stable_release': 'luminous', 'cluster_address': '192.168.4.200', 'cluster_name': u'test', ...}
host = <testinfra.host.Host object at 0x7f75bea4d090>

    def test_osd_services_are_running(self, node, host):
        # TODO: figure out way to paramaterize node['osds'] for this test
        for osd in node["osds"]:
>           assert host.service("ceph-osd@%s" % osd).is_running
E           AssertionError: assert False
E            +  where False = <service ceph-osd@>.is_running
E            +    where <service ceph-osd@> = <class 'testinfra.modules.base.SystemdService'>(('ceph-osd@%s' % ''))
E            +      where <class 'testinfra.modules.base.SystemdService'> = <testinfra.host.Host object at 0x7f75bea4d090>.service

../../../tests/osd/test_osds.py:23: AssertionError
________________ TestOSDs.test_osd_are_mounted[ansible://osd0] _________________
[gw3] linux2 -- Python 2.7.13 /tmp/tox.HsMm8N7kan/luminous-ansible2.3-journal_collocation_auto/bin/python2

self = <tests.osd.test_osds.TestOSDs object at 0x7fd856293b90>
node = {'address': '192.168.3.100', 'ceph_stable_release': 'luminous', 'cluster_address': '192.168.4.200', 'cluster_name': u'test', ...}
host = <testinfra.host.Host object at 0x7fd854f522d0>

    @pytest.mark.no_docker
    def test_osd_are_mounted(self, node, host):
        # TODO: figure out way to paramaterize node['osd_ids'] for this test
        for osd_id in node["osd_ids"]:
            osd_path = "/var/lib/ceph/osd/{cluster}-{osd_id}".format(
                cluster=node["cluster_name"],
                osd_id=osd_id,
            )
>           assert host.mount_point(osd_path).exists
E           assert False
E            +  where False = <[KeyError("u'device'") raised in repr()] SafeRepr object at 0x7fd8562b8a70>.exists
E            +    where <[KeyError("u'device'") raised in repr()] SafeRepr object at 0x7fd8562b8a70> = <class 'testinfra.modules.base.LinuxMountPoint'>('/var/lib/ceph/osd/test-')
E            +      where <class 'testinfra.modules.base.LinuxMountPoint'> = <testinfra.host.Host object at 0x7fd854f522d0>.mount_point

../../../tests/osd/test_osds.py:38: AssertionError
____________ TestOSDs.test_osd_services_are_enabled[ansible://osd0] ____________
[gw1] linux2 -- Python 2.7.13 /tmp/tox.HsMm8N7kan/luminous-ansible2.3-journal_collocation_auto/bin/python2

self = <tests.osd.test_osds.TestOSDs object at 0x7f75bfd73cd0>
node = {'address': '192.168.3.100', 'ceph_stable_release': 'luminous', 'cluster_address': '192.168.4.200', 'cluster_name': u'test', ...}
host = <testinfra.host.Host object at 0x7f75bea4d090>

    def test_osd_services_are_enabled(self, node, host):
        # TODO: figure out way to paramaterize node['osds'] for this test
        for osd in node["osds"]:
>           assert host.service("ceph-osd@%s" % osd).is_enabled
E           AssertionError: assert False
E            +  where False = <service ceph-osd@>.is_enabled
E            +    where <service ceph-osd@> = <class 'testinfra.modules.base.SystemdService'>(('ceph-osd@%s' % ''))
E            +      where <class 'testinfra.modules.base.SystemdService'> = <testinfra.host.Host object at 0x7f75bea4d090>.service
```

Thanks!

**Comment by leseb** [comment]
@andrewschoen forget about my last comment, I understand what's wrong.

**Comment by leseb** [comment]
@andrewschoen I think the problem is that testinfra can not get the fact `devices`. Thus putting all the values at 0. Any idea how to sort this out?

**Comment by leseb** [comment]
@andrewschoen I think the problem is that testinfra can not get the fact `devices`. Thus putting all the values at 0. Any idea how to sort this out?

**Comment by leseb** [comment]
@andrewschoen I think the problem is that testinfra can not get the fact `devices`. Thus putting all the values at 0. Any idea how to sort this out?

**Comment by andrewschoen** [comment]
@leseb ``node["vars"]["devices"]`` is actually looking for a value for ``devices`` defined in ``group_vars`` for the testing scenario. Your new osd autodiscovery testing scenarios don't have ``devices`` defined so these tests fail. You need to modify these tests so that they work the same for scenarios that do not define ``devices`` or you need to mark them in such a way so that they are not run for scenarios that they are not valid for. We do something similar for tests that do not apply to containerized clusters.

See: https://github.com/ceph/ceph-ansible/blob/master/tests/conftest.py#L29-L30 and https://github.com/ceph/ceph-ansible/blob/master/tests/functional/tests/test_install.py#L18

**Comment by leseb** [comment]
@andrewschoen thanks, in the meantime I've tried that: https://github.com/ceph/ceph-ansible/pull/1803/files#diff-484462fced51d1a06b1d93b4a44dd535R50

**Comment by andrewschoen** [comment]
> @andrewschoen thanks, in the meantime I've tried that: https://github.com/ceph/ceph-ansible/pull/1803/files#diff-484462fced51d1a06b1d93b4a44dd535R50

@leseb this will break the lvm scenario though, I think what we should be checking the scenario name there and setting it appropriately per scenario. Also, you're still not changing the correct ``devices``. The ``num_devices`` variable is used in tests, but it is not the save as ``node['vars']['devices']``.

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-xenial_cluster

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-update_dmcrypt

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-update_cluster

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-switch_to_containers

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-shrink_osd

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-shrink_mon

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-lvm_osds

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-journal_collocation_auto_dmcrypt

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-bluestore_docker_cluster

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-bluestore_cluster

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-journal_collocation_auto

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-journal_collocation_auto_dmcrypt

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-journal_collocation_auto

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-bluestore_dmcrypt_journal_collocation

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-bluestore_docker_dmcrypt_journal_collocation

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-bluestore_journal_collocation

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-centos7_cluster

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-dmcrypt_journal

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-dmcrypt_journal_collocation

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-journal_collocation_auto_dmcrypt

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-update_cluster

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-update_dmcrypt

**Comment by leseb** [comment]
jenkins test this please

**Comment by leseb** [comment]
jenkins test luminous-ansible2.3-journal_collocation_auto

**Comment by leseb** [comment]
jenkins test this please

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-bluestore_cluster

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-bluestore_dmcrypt_journal

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-bluestore_dmcrypt_journal_collocation

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-bluestore_docker_dmcrypt_journal_collocation

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-dmcrypt_journal

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-dmcrypt_journal_collocation

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-journal_collocation_auto

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-update_dmcrypt

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-journal_collocation_auto_dmcrypt

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-purge_dmcrypt

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-bluestore_cluster

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-dmcrypt_journal_collocation

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-docker_cluster

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-docker_dedicated_journal

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-bluestore_dmcrypt_journal_collocation

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-bluestore_docker_dmcrypt_journal_collocation

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-bluestore_docker_dmcrypt_journal_collocation

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-purge_docker_cluster

**Comment by guits** [comment]
jenkins test luminous-ansible2.3-purge_docker_cluster

**Pull request merged**